com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js

```diff
@@ -1,247 +1,333 @@
+        if (currentAttempt == null) currentAttempt = 0;
+
+        var xhr = new XMLHttpRequest();
+        xhr.open('GET', url, true);
+        xhr.responseType = options.responseType;
+        xhr.onprogress = function(e) {
+            if (e.lengthComputable) {
+                onprogress(xhr, e.loaded, e.total);
+                return;
+            }
+            var total = xhr.getResponseHeader('content-length')
+            var encoding = xhr.getResponseHeader('content-encoding')
+            if (total && encoding && (encoding.indexOf('gzip') > -1)) {
+                total /= 0.4; // assuming 40% compression rate
+                onprogress(xhr, e.loaded, total);
+            } else {
+                onprogress(xhr, 1, 1);
+            }
+        };
+        xhr.onload = function(e) {
+            if (xhr.readyState === 4) {
+                if (xhr.status === 200) {
+                    onload(xhr, e);
+                } else {
+                    onerror(xhr, e);
+                }
+            }
+        };
+        xhr.onerror = function(e) {
+            if (currentAttempt < options.retryCount) {
+                setTimeout(function() {
+                    FileLoader.load(url, onprogress, onloaded, onerror, options, currentAttempt + 1);
+                }, options.retryInterval);
+            } else {
+                onerror(xhr, e);
+            }
+        };
+        xhr.send(null);
+    }
+};
+
+
+var EngineLoader = {
+    loadScriptAsync: function(src, fromProgress, toProgress) {
+        FileLoader.load(
+            src,
+            function (xhr, loaded, total) {
+                var progress = fromProgress + (loaded / total) * (toProgress - fromProgress);
+                Progress.updateProgress(progress);
+            },
+            function (xhr, e) {
+                var tag = document.createElement("script");
+                tag.text = xhr.response;
+                document.head.appendChild(tag);
+            },
+            function (xhr, e) {
+                throw "Async engine load error " + src;
+            }
+        );
+    },
+
+    // load engine (asm.js or wasm.js + wasm)
+    // engine load progress goes from 1-50% for ams.js
+    // engine load progress gors from 1-10% for wasm.js and 10-50% for .wasm
+    load: function(appCanvasId, exeName) {
+        Progress.addProgress(Module.setupCanvas(appCanvasId));
+        if (Module['isWASMSupported']) {
+            EngineLoader.loadScriptAsync(exeName + '_wasm.js', 0, 10);
+        } else {
+            EngineLoader.loadScriptAsync(exeName + '_asmjs.js', 0, 50);
+        }
+    }
+}
+
+
 /* ********************************************************************* */
-/* Load and combine data that is split into archives                     */
+/* Load and combine game archive data that is split into archives        */
 /* ********************************************************************* */
 
-var Combine = {
```

Comment:
Combine was not a very accurate name. The purpose of Combine.* functions was mainly to download the game archive files (and combine them if a file consisted of more than one piece).